### ISSUE TITLE
[Embed block] Paragraph block does not detect embeddable link that is pasted from Gboard

### ISSUE THREAD

**Issue by jd-alexander** [wordpress-mobile/gutenberg-mobile#4203]
**Describe the bug**

When the Gboard, Google Keyboard is utilized as the default Keyboard on Android and the clipboard content is channeled via Google Keyboard's clipboard the Paragraph does not detect that the content is being pasted, therefore the Embed block Bottom Sheet is not triggered. 

**To Reproduce**

1. Ensure Google Keyboard is set as your default keyboard.
2. Copy an embeddable link to the device's clipboard.
3. Open the app and create a new post.
4. Notice that the embeddable link is on the Google Keyboard's UI. 
5. Press the embeddable link. 
6. Notice the Paragraph is populated with the link but the Embed Bottom Sheet is not displayed. 

**Expected behavior**

The Embed Bottom Sheet should be shown when link-based content from the clipboard is pasted into a Paragraph block.

**Screenshots**

https://user-images.githubusercontent.com/1509205/140267078-e33c10da-1d15-4c31-9f76-667699a0273d.mov

**Smartphone (please complete the following information):**
 - Device: Google Pixel 4
 - OS: Android
 - Version 11

**Comment by jd-alexander** [comment]
Based on my investigation this issue occurs because the clipboard content that is shown to the user as a suggestion on Gboard is actually typed in as text, and is not actually pasted. I verified this by utilizing the following troubleshooting steps below. 

1. I put a `console.log` within the `onSelectionChange` of the `rich-text` component. ([function reference](https://github.com/WordPress/gutenberg/blob/trunk/packages/rich-text/src/component/index.native.js#L581))
2. I also put a `console.log` within the `onPaste` of the `rich-text` component. ([function reference](https://github.com/WordPress/gutenberg/blob/trunk/packages/rich-text/src/component/index.native.js#L501))
3. After setting up the logging, I noticed that when I long pressed on the Paragraph block and did the standard Android paste action, the `onPaste` function was triggered. 
4. However, if Gboard's suggestion of the pasted link is clicked then only `onSelectionChange` is triggered.

**Comment by jd-alexander** [comment]
@fluiddot I want to hear your thoughts on a potential solution for this issue. Thanks in advance 🙇🏾 

## Potential ideas 

### Preferred solution

1. **Introduce Gboard detection logic within the RichText component** - We could modify the `oonSelectionChangeFromAztec` function ([reference](https://github.com/WordPress/gutenberg/blob/f64da2b69c789ec6b62532ba43c814dc5c2bb9c0/packages/rich-text/src/component/index.native.js#L627)) and detect that the selection change that was received is `Gboard` related. Once this is ascertained, we could then propagate this `Selection Change` to the `onPaste` handler with the event object it accepts. 

### Other potential solutions and considerations

2. **Send keyboard name over the bridge** - Since this is affecting `Gboard`'s there's a mechanism for fetching the name of the keyboard via the [InputMethodManager](https://developer.android.com/reference/android/view/inputmethod/InputMethodManager#getEnabledInputMethodList()). This would need to be sent over the bridge so it might not be ideal.

 3. **Modify Aztec Android's onSelectionChange** - I was thinking that it could be resolved by modifying `Aztec Android` to detect when the `Gboard` paste takes place based on the behavior of the `onSelectionChange` i.e since its a longer input being entered than one letter at a time, we should be able to write logic to detect this change and trigger the `onPaste`. I am looking at https://github.com/WordPress/gutenberg/pull/32471 for inspiration. The mechanism for achieving this would be similar to a [Clipboard event in Android](https://developer.android.com/reference/android/content/ClipboardManager#setPrimaryClip(android.content.ClipData)) and send this data via the `ReactAztec` `onPaste` [mechanism](https://github.com/WordPress/gutenberg/blob/f64da2b69c789ec6b62532ba43c814dc5c2bb9c0/packages/react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecText.java#L487).
 
 ## Prioritization
 
Before an attempt is made on this issue, I am also wondering if this bug is one that should be prioritized for a fix. This thought came to my mind because the user can still long-press on the Paragraph block and utilize the normal Paste functionality instead of the keyboard itself. This solution would be introducing custom behavior for a Keyboard. Nonetheless, we have done this before for other issues, therefore it is worth consideration.

**Comment by hypest** [comment]
👋 I had a super quick look at the case where GBoard inserts the copied text as text change event (instead of Paste event) and I think we can probably tap into the [onTextUpdate()](https://github.com/WordPress/gutenberg/blob/trunk/packages/rich-text/src/component/index.native.js#L301-L321) function and detect the URL from there. The only "fear" I have is of a case where GBoard won't insert the text as a whole but break it down to separate text insertion events for _reasons_.

I went ahead and tried using a modified version of the [onPaste function](https://github.com/WordPress/gutenberg/blob/trunk/packages/rich-text/src/component/index.native.js#L501) (to pick up the `text` property from the `nativeEvent`) which kinda worked. It detected the URL but I didn't see the bottomsheet popup. I feel there's some potential there though.

**Comment by fluiddot** [comment]
I'd like to highlight that this issue also affects other pasting features like creating blocks when pasting HTML, so I'd lean towards finding a way to detect when the Gboard is being used and somehow detect when the content is pasted.

Not sure if it will work but maybe we could check in the `onTextChanged` function ([reference](https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecManager.java#L641-L693)) if the text change matches the content from the clipboard (in case Gboard exposes this data via the [`ClipboardManager`](https://developer.android.com/reference/android/content/ClipboardManager)) and the Gboard is being used.

**Comment by fluiddot** [comment]
> Before an attempt is made on this issue, I am also wondering if this bug is one that should be prioritized for a fix. This thought came to my mind because the user can still long-press on the Paragraph block and utilize the normal Paste functionality instead of the keyboard itself. This solution would be introducing custom behavior for a Keyboard. Nonetheless, we have done this before for other issues, therefore it is worth consideration.

Good point, to be honest, I'm not sure how critical we should consider the issue but, as I commented in the [previous comment](https://github.com/wordpress-mobile/gutenberg-mobile/issues/4203#issuecomment-961808844), all the functionality we do when pasting content in the `RichText` component is not executed when pasting from the Gboard keyboard, not just the one for the embed block. Not sure how many users use the Gboard (or other keyboards that might have the same issue) to consider this high priority 🤔 .

**Comment by jd-alexander** [comment]
> 👋 I had a super quick look at the case where GBoard inserts the copied text as text change event (instead of Paste event) and I think we can probably tap into the onTextUpdate() function and detect the URL from there. The only "fear" I have is of a case where GBoard won't insert the text as a whole but break it down to separate text insertion events for reasons.
I went ahead and tried using a modified version of the onPaste function (to pick up the text property from the nativeEvent) which kinda worked. It detected the URL but I didn't see the bottomsheet popup. I feel there's some potential there though.

Thanks for chiming in 🙇🏾 I agree that the text being broken down into separate insertion events would be something to look out for as it would impact the ability of the logic to pickup URLs.  

> Good point, to be honest, I'm not sure how critical we should consider the issue but, as I commented in the previous comment, all the functionality we do when pasting content in the RichText component is not executed when pasting from the Gboard keyboard, not just the one for the embed block. Not sure how many users use the Gboard (or other keyboards that might have the same issue) to consider this high priority 🤔 .

Agreed. I will put a low priority label for now since the long-press paste functionality works as expected and would be the ideal channel to trigger the paste behavior. If we get more reports of this impacting users then we can combine some of the discussed ideas to create a solution.
